codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java

```diff
@@ -59,234 +58,27 @@
-                name.forEachByte(new ValidateNameProcessor(name));
-            } catch (Throwable t) {
-                PlatformDependent.throwException(t);
-            }
-        }
-
-        private static void validateName(CharSequence name) {
-            // Go through each characters in the name.
-            for (int index = 0; index < name.length(); index++) {
-                char c = name.charAt(index);
-
-                // Check to see if the character is not an ASCII character.
-                if (c > 127) {
-                    throw new IllegalArgumentException("a header name cannot contain non-ASCII characters: " + name);
-                }
-
-                // Check for prohibited characters.
-                validateNameChar(c, name);
-            }
-        }
-
-        private static void validateNameChar(int character, CharSequence seq) {
-            if ((character & HIGHEST_INVALID_NAME_CHAR_MASK) == 0 && LOOKUP_TABLE[character] != 0) {
-                throw new IllegalArgumentException(
-                        "a header name cannot contain the following prohibited characters: =,;: \\t\\r\\n\\v\\f: " +
-                                seq);
-            }
-        }
-    }
-
-    private static final HttpHeadersValidationConverter
-        VALIDATE_OBJECT_CONVERTER = new HttpHeadersValidationConverter(true);
-    private static final HttpHeadersValidationConverter
-        NO_VALIDATE_OBJECT_CONVERTER = new HttpHeadersValidationConverter(false);
-    private static final HttpHeadersNameConverter VALIDATE_NAME_CONVERTER = new HttpHeadersNameConverter(true);
-    private static final HttpHeadersNameConverter NO_VALIDATE_NAME_CONVERTER = new HttpHeadersNameConverter(false);
 
     public DefaultHttpHeaders() {
         this(true);
     }
 
     public DefaultHttpHeaders(boolean validate) {
-        this(true, validate? VALIDATE_NAME_CONVERTER : NO_VALIDATE_NAME_CONVERTER, false);
+        this(validate, false);
     }
 
-    public DefaultHttpHeaders(boolean validate, boolean singleHeaderFields) {
-        this(true, validate? VALIDATE_NAME_CONVERTER : NO_VALIDATE_NAME_CONVERTER, singleHeaderFields);
+    protected DefaultHttpHeaders(boolean validate, boolean singleHeaderFields) {
+        this(true, validate ? HeaderNameValidator.INSTANCE : DefaultTextHeaders.NO_NAME_VALIDATOR, singleHeaderFields);
     }
 
-    protected DefaultHttpHeaders(boolean validate, NameConverter<CharSequence> nameConverter,
+    protected DefaultHttpHeaders(boolean validate,
+                                 DefaultHeaders.NameValidator<CharSequence> nameValidator,
                                  boolean singleHeaderFields) {
-        headers = new DefaultTextHeaders(true,
-                validate ? VALIDATE_OBJECT_CONVERTER : NO_VALIDATE_OBJECT_CONVERTER, nameConverter, singleHeaderFields);
+        headers = new DefaultTextHeaders(
+                    new TreeMap<CharSequence, Object>(AsciiString.CHARSEQUENCE_CASE_INSENSITIVE_ORDER),
```

Comment:
@normanmaurer - @buchgr could answer more. But I think it had to do with:

1. not having to "resize/copy" 
2. not having to generate hash codes and using the equals method continuously.